// verification-code-react/lib/verificationcode.js
import React, {Component} from "react"
import PropTypes from "prop-types"
import "./verificationcode.css"

export default class VerificationCode extends Component {
	constructor(props) {
		super(props)

		this.startTime = props.startTime
		this.holdString = props.autoStart ? props.startTime + "s" : props.holdString
		this.timer = null
		this.isRuning = false
		this.className = props.autoStart ? "container-countdown" : "container-countdown stop"

		this.state = {
			currentTime: props.startTime,
			holdString: this.holdString
		}
	}

	componentWillReceiveProps(nextProps) {
		if(nextProps.restart) {
			clearInterval(this.timer)
			this.isRuning = false
			this.className = "container-countdown stop"
			this.setState({
				holdString: this.holdString,
				currentTime: this.startTime
			})
		}
	}

	componentDidMount() {
		if(this.props.autoStart) this.restart(false)
	}

	restart(isFromClick) {
		if(this.isRuning) {
			return
		} else {
			//do anything in a loop just one time...
			this.props.callback()
		}

		//click should change value intime
		if(isFromClick) {
			this.setState({
				currentTime: this.startTime,
				holdString: null
			})
			this.className = "container-countdown"
		}

		this.isRuning = true
		this.timer = setInterval(() => {
			let _currentTime = this.state.currentTime
			let _state = null

			if(_currentTime > 1) {
				_state = {
					currentTime: _currentTime - 1,
					holdString: null
				}
			} else {
				_state = {
					holdString: this.holdString,
					currentTime: this.startTime
				}
				clearInterval(this.timer)
				this.isRuning = false
				this.className = "container-countdown stop"
			}

			this.setState(_state)
		}, 1000)
	}

	render() {
		let _currentTime = this.state.currentTime
		let _holdString = this.state.holdString
		let _defaultValue = _holdString ? _holdString : _currentTime + "s"

		let _restart = this.restart.bind(this, true)
		
		return(
			<input
				className={this.className}
				onClick={_restart}
				type="button"
				defaultValue= {_defaultValue} />
		)
	}
}

VerificationCode.propTypes = {
	startTime: PropTypes.number.isRequired,
	holdString: PropTypes.string.isRequired,
	callback: PropTypes.func.isRequired,
	autoStart: PropTypes.bool.isRequired
}
